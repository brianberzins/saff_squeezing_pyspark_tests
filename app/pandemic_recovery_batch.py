from datetime import datetime, date
from typing import List

from pyspark.sql import functions as F, DataFrame
from pyspark.sql.functions import udf
from pyspark.sql.types import IntegerType


def transform(business_df: DataFrame,
              checkin_df: DataFrame,
              browser_reviews_df: DataFrame,
              tips_df: DataFrame,
              mobile_reviews_df: DataFrame,
              run_date: datetime = datetime.today()):

    checkin_df = checkin_df.withColumn("checkins_list", F.split(checkin_df.date, ","))
    checkin_df = checkin_df.withColumn("date", F.explode(checkin_df.checkins_list))
    checkin_df = checkin_df.withColumn("date", F.trim(checkin_df.date))
    checkin_df = checkin_df.select(
        F.col("business_id"),
        F.col("user_id"),
        F.col("date")
    )


    reviews_df = count_reviews(checkin_df, mobile_reviews_df, browser_reviews_df, run_date)
    checkin_df = count_checkins(checkin_df, run_date)

    tips_df = count_tips(tips_df, run_date)

    return construct_post_pandemic_recovery_df(
        business_df, checkin_df, reviews_df, run_date, tips_df
    )


def construct_post_pandemic_recovery_df(business_df, checkin_df, reviews_df, run_date, tips_df):
    pandemic_recovery_df = business_df.join(checkin_df, on="business_id", how='left').fillna(0)
    pandemic_recovery_df = pandemic_recovery_df.join(reviews_df, on="business_id", how='left').fillna(0)
    pandemic_recovery_df = pandemic_recovery_df.join(tips_df, on="business_id", how='left').fillna(0)
    pandemic_recovery_df = pandemic_recovery_df.withColumn("num_interactions",
                                                           pandemic_recovery_df.num_reviews +
                                                           pandemic_recovery_df.num_tips +
                                                           pandemic_recovery_df.num_checkins)
    pandemic_recovery_df = pandemic_recovery_df.withColumn("dt", F.lit(run_date.strftime("%Y-%m-%d")))
    pandemic_recovery_df = pandemic_recovery_df.select(
        "business_id", "name", "num_tips", "num_checkins", "num_reviews", "num_interactions", "dt"
    )
    return pandemic_recovery_df


def count_tips(tips_df, run_date):
    tips_df = tips_df.filter(tips_df.date == run_date.date())
    tips_df = tips_df.groupby("business_id").count()
    tips_df = tips_df.withColumnRenamed("count", "num_tips")
    return tips_df


def count_checkins(checkin_df, run_date):
    checkin_df = checkin_df.filter(checkin_df.date == run_date.date())
    checkin_df = checkin_df.groupby("business_id").count()
    checkin_df = checkin_df.withColumnRenamed("count", "num_checkins")
    return checkin_df


def count_reviews(checkin_df, mobile_reviews_df, browser_reviews_df, run_date):
    mobile_reviews_df = mobile_reviews_df.join(checkin_df, on=['business_id','date', 'user_id'], how="left_anti").select(*mobile_reviews_df.columns)
    reviews_df = mobile_reviews_df.union(browser_reviews_df)
    reviews_df = reviews_df.filter(reviews_df.date == run_date.date())
    reviews_df = reviews_df.groupby("business_id").count()
    reviews_df = reviews_df.withColumnRenamed("count", "num_reviews")
    return reviews_df


