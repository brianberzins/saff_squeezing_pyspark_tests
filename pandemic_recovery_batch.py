from datetime import datetime
from typing import List

from pyspark.sql import functions as F, DataFrame
from pyspark.sql.functions import udf
from pyspark.sql.types import IntegerType


def transform(business_df: DataFrame,
              checkin_df: DataFrame,
              browser_reviews_df: DataFrame,
              tips_df: DataFrame,
              mobile_reviews_df: DataFrame,
              run_date: datetime = datetime.today()):

    checkin_df = checkin_df.withColumn("checkins_list", F.split(checkin_df.date, ","))
    checkin_df = checkin_df.select(F.col("business_id"), F.explode(F.col("checkins_list")).alias("date"))

    reviews_df = count_reviews(checkin_df, mobile_reviews_df, browser_reviews_df, run_date)
    checkin_df = count_checkins(checkin_df, run_date)

    tips_df = count_tips(tips_df, run_date)

    return construct_post_pandemic_recovery_df(
        business_df, checkin_df, reviews_df, run_date, tips_df
    )


def construct_post_pandemic_recovery_df(business_df, checkin_df, reviews_df, run_date, tips_df):
    entity_with_activity_df = business_df.join(checkin_df, on="business_id", how='left').fillna(0)
    entity_with_activity_df = entity_with_activity_df.join(reviews_df, on="business_id", how='left').fillna(0)
    entity_with_activity_df = entity_with_activity_df.join(tips_df, on="business_id", how='left').fillna(0)
    entity_with_activity_df = entity_with_activity_df.withColumn("num_interactions",
                                                                 entity_with_activity_df.num_reviews +
                                                                 entity_with_activity_df.num_tips +
                                                                 entity_with_activity_df.num_checkins)
    entity_with_activity_df = entity_with_activity_df.withColumn("dt", F.lit(run_date))
    entity_with_activity_df = entity_with_activity_df.select(
        "business_id", "name", "num_tips", "num_checkins", "num_reviews", "num_interactions", "dt"
    )
    return entity_with_activity_df


def count_tips(tips_df, run_date):
    tips_df = tips_df.filter(tips_df.date == run_date.date())
    tips_df = tips_df.groupby("business_id").count()
    tips_df = tips_df.withColumnRenamed("count", "num_tips")
    return tips_df


def count_checkins(checkin_df, run_date):
    checkin_df = checkin_df.filter(checkin_df.date == run_date.date())
    checkin_df = checkin_df.groupby("business_id").count()
    checkin_df = checkin_df.withColumnRenamed("count", "num_checkins")
    return checkin_df


def count_reviews(checkin_df, mobile_reviews_df, browser_reviews_df, run_date):
    mobile_reviews_df = mobile_reviews_df.join(checkin_df, on=['business_id'], how="left_anti").select(mobile_reviews_df.columns)
    reviews_df = mobile_reviews_df.union(browser_reviews_df)
    reviews_df = reviews_df.filter(reviews_df.date == run_date.date())
    reviews_df = reviews_df.groupby("business_id").count()
    reviews_df = reviews_df.withColumnRenamed("count", "num_reviews")
    return reviews_df


