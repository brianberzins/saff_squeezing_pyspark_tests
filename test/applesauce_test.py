from datetime import datetime


from chispa import assert_df_equality
from pyspark.sql.types import StructType, StructField, StringType, LongType

from pandemic_recovery_batch import count_interactions_from_reviews


def create_df(spark, schema, data):
    return spark.createDataFrame(schema=schema, data=data)


SCHEMA = StructType([
            StructField('user_id', StringType()),
            StructField('business_id', StringType()),
            StructField('date', StringType())])

SCHEMA2 = StructType([
            StructField('business_id', StringType()),
            StructField('num_reviews', LongType(), False)])

def test_keeps_mobile_reviews_without_checkins(spark):
    mobile_df = create_df(spark, SCHEMA, [{'user_id': 'uid', 'business_id': 'bid', 'date': '2022-04-14'}])
    __ = create_df(spark, SCHEMA, [])

    reviews_df = count_interactions_from_reviews(__, mobile_df, __, datetime(2022, 4, 14))

    expected_df = create_df(spark, SCHEMA2, [{ 'business_id': 'bid', 'num_reviews': 1}])
    assert_df_equality(reviews_df, expected_df)

def test_does_not_count_mobile_reviews_with_checkins(spark):
    mobile_df = create_df(spark, SCHEMA, [{ 'business_id': 'bid', 'user_id': 'uid', 'date': '2022-04-14'}])
    checkin_df = create_df(spark, SCHEMA, [{'user_id': 'uid', 'business_id': 'bid', 'date': '2022-04-14'}])
    __ = create_df(spark, SCHEMA, [])

    reviews_df = count_interactions_from_reviews(checkin_df, mobile_df, __, datetime(2022, 4, 14))

    expected_df = create_df(spark, SCHEMA2, [])
    assert_df_equality(reviews_df, expected_df)