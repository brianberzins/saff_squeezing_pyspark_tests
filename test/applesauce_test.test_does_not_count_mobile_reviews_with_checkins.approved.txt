+-----------+-----------+
|business_id|num_reviews|
+-----------+-----------+
+-----------+-----------+
